### ISSUE TITLE
Currency and selected bank account for shops, real estate desk, etc. revert to defaults on server restarts

### ISSUE THREAD

**Issue by D3nnis3n** [StrangeLoopGames/EcoIssues#17463]
Really, we had that issue several times on 8.x already. Can we get this fixed PERMANENTLY?

**Comment by D3nnis3n** [comment]
Credits might already be fixed - double check if bank accounts are fixed as well!

**Comment by beastyn** [comment]
Bank account was switched after server restart in Real Estate.

**Comment by gzatravkin** [comment]
Repro steps or some ideas?

Tried to reproduce locally by:
spawn store
money addbankaccounts 

Bank account and currency survived server restart
and set some not default bank account.
It remains after server restart

**Comment by IcedForge** [comment]
only the initial person changing settings in the store would take effect
So if 2 ppl access the store with edit permissions, only first player interacting would take effect
f.ex

Player 1 opens store, player 2 opens store, then player 2 makes a change after player 1 left
the change player 2 makes, will revert back
because only the initial auth opening of a window will actualize.

**Comment by IcedForge** [comment]
Which i might add a different issue BUT its causing the same behavior type so i think some people might be having conflicted ideas on what is causing it.

**Comment by gzatravkin** [comment]
I investigated this issue. 
There are simmiliar issue (which is already fixed but stil not in playtest): https://github.com/StrangeLoopGames/EcoIssues/issues/17194 . I assume that almost all problems with switching of accounts/currencies are solved with fix for it. 

"Bank account was switched after server restart in Real Estate."
Am I right it was about this one? 
![image](https://user-images.githubusercontent.com/53317567/89947702-46d72300-dbfb-11ea-8bd6-f907cf74164c.png)
@beastyn 

If so, I believe it's ok if it's switched. It have no "long time" value like bank account in shop. It will be switched on client reload too.

It's just current selected choise to 
1. Create new sale (which should keep this account even after server restart)
2. Buy some deeds (and in this case it's just preferred method of payment. It will be saved while UI is cached).

 This bank account is not part of real estate desk state, it's prefs of current player (and may be different for each player).

So if I understood it right, let's close this issue and keep track of switching in other places (shops for example) here https://github.com/StrangeLoopGames/EcoIssues/issues/17194

**Comment by D3nnis3n** [comment]
No, i mean the bank account of the seller for the real estate desk, where the money of sold deeds goes into. Von: gzatravkinGesendet: Dienstag, 11. August 2020 23:00An: StrangeLoopGames/EcoIssuesCc: Dennis Scholz; AuthorBetreff: Re: [StrangeLoopGames/EcoIssues] Currency and selected bank account for shops, real estate desk, etc. revert to defaults on server restarts (#17463) I investigated this issue.There are simmiliar issue (which is already fixed but stil not in playtest): #17194 . I assume that almost all problems with switching of accounts/currencies are solved with fix for it."Bank account was switched after server restart in Real Estate."Am I right it was about this one?@beastynIf so, I believe it's ok if it's switched. It have no "long time" value like bank account in shop. It will be switched on client reload too.It's just current selected choise toCreate new sale (which should keep this account even after server restart)Buy some deeds (and in this case it's just preferred method of payment. It will be saved while UI is cached).This bank account is not part of real estate desk state, it's prefs of current player (and may be different for each player).So if I understood it right, let's close this issue and keep track of switching in other places (shops for example) here #17194—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by beastyn** [comment]
@gzatravkin yeap I am talking about this field and this filed should save it's value as it used by the seller and the buyer.
If you are selling this account will be used for receiving money.
If you are buying it will be used for money spending.

**Comment by gzatravkin** [comment]
@beastyn at least acording the code, it's the account what will be assigned to deed sale. 
I.e. you set ac1, create deed sale 1. Set ac2, create deed sale 2. So after sale 1 you will have money at ac1, and after sale 2 at ac2.
Is it working differently? 

For buying yes, it's like payment method (which now you have to select at the moment of buying). 

Shouldn't be a problem to add pref value anyway.

**Comment by gzatravkin** [comment]
I added stored pref value to dropdown, but it's just cosmetic change. As I see, the problem is that interface is unclear (users think that they will recieve money in last selected bank account but it's bank account to assign to new deed sale at moment of creation).

@beastyn please confirm. Maybe we could have another issue to some rework on that interface if it's the case.

**Comment by D3nnis3n** [comment]
> but it's bank account to assign to new deed sale at moment of creation).

Nope, that can't be true.

All the 40 Deeds we sell on White-Tiger were set up one after the other via my account. While the first few put the money into the correct account, after server restart - when that field resetted - money was put into my personal account instead. So there must be a logic issue as well.

**Comment by beastyn** [comment]
@gzatravkin I'm not sure about code side, you can ask @mirasrael as it was his realization.
But I can show what I see,

1. Person 1 make a sale, place his new bank account for receiving the payment (its not default)
2. Person two wants to buy something on this desk using not a default account
![bandicam 2020-08-18 14-51-06-695](https://user-images.githubusercontent.com/27898520/90498919-5528c000-e162-11ea-8b02-251dcd0987e6.jpg)

3. Dealing
Person 1 has got money on its account from desk
![bandicam 2020-08-18 14-53-32-915](https://user-images.githubusercontent.com/27898520/90499272-c36d8280-e162-11ea-82ce-8e49eb87b94c.jpg)

Person 2 has spent money from account which was on a desk
![bandicam 2020-08-18 14-55-30-686](https://user-images.githubusercontent.com/27898520/90499360-e4ce6e80-e162-11ea-9e05-1147f69e14e1.jpg)

So yes, its important for the both sides and should be consistent

**Comment by mirasrael** [comment]
It saves listing data including who posted, which account used and which currency selected. So it shouldn't be affected by Credit Component selection unless it replaces listing selection somehow.

**Comment by mirasrael** [comment]
I think we should push this issue for post-9.0, because as Grigory said it is player preference which isn't persisted in the component. If we wanna save it then it should be server-side setting, but in this case it either should be same account for all tables (which also may be not a solution you expect, because you may want to use one account for Store and another account for Real Estate desk and probably another one for another Store etc). So probably in this case you expect it to be per table setting, but it will require to extend these components with persistent data which isn't easy fix. And may cause additional bugs and require additional improvements.
If we have the issue with listed deeds receiving account reset after server restart though then it should be fixed. So need to check this part. But from code I checked it should persist target account and currency when deed created.

**Comment by gzatravkin** [comment]
@beastyn can you confirm if we have issue with resetting target account?

I.e. after create deed sale with setted account 2 and restart of server user will recieve money in default accoutn instead of account 2

**Comment by beastyn** [comment]
Soo, what I see now.

1. Account for a SALLER is assigned to the deed and won't be changed after server restart,
@D3nnis3n if you are creating a sale it will remember account from UI for this deed. So you can create deed 1 sale with account 1, deed 2 sale with account 2.
Problem is - you can't understand it from UI and you won't see in the sale list which account have you used for sale.
Maybe for the fast fix we can add a column with account used for sale (to be seen by seller)?
2.  Now account won't be saved on a desk at all, that will be cleared to default with ui cache clearing. (use /toggleuicache)
Not so critical after this account logic changes, but still frustrating a bit (it was fine before this fix.)

So, from my side I can recommend as fast fix
- add column for seller to see account used for money receiving
- save account for the session at least
- maybe add some instruction somewhere (like in tooltip).

Only if it won't need any big changes.
(can be decided after playtest)

**Comment by mirasrael** [comment]
It works in 8.3.3 same way (it remembers account for listing). So I don't see why it so urgent to fix now if it worked this way for 1.5 years and we didn't hot fix it. 
I already said why saving account even for session doesn't look like a trivial solution. Should it be remembered for all tables? Only for real estate desks if we have multiple on server? Save in session for every real estate desk? Not only for real estate desk? So it will require both design and implementation effort. I suggest to push this change and decision for later.

I think as fast fix it will be enough to add info about account to the tooltip (probably this infor should only be shown to Deed owner).

**Comment by D3nnis3n** [comment]
Once again, we didn’t have this issue in 8.3.3. Nothing more i could say. The first time we encountered this was on 9.0. Von: Alexander BondarevGesendet: Mittwoch, 19. August 2020 16:33An: StrangeLoopGames/EcoIssuesCc: Dennis Scholz; MentionBetreff: Re: [StrangeLoopGames/EcoIssues] Currency and selected bank account for shops, real estate desk, etc. revert to defaults on server restarts (#17463) It works in 8.3.3 same way (it remembers account for listing). So I don't see why it so urgent to fix now if it worked this way for 1.5 years and we didn't hot fix it.I already said why saving account even for session doesn't look like a trivial solution. Should it be remembered for all tables? Only for real estate desks if we have multiple on server? Save in session for every real estate desk? Not only for real estate desk? So it will require both design and implementation effort. I suggest to push this change and decision for later.I think as fast fix it will be enough to add info about account to the tooltip (probably this infor should only be shown to Deed owner).—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by beastyn** [comment]
@D3nnis3n It will put money into your account which was used when creating a deed even after server restart. The only problem for now is that there is no any info for seller about it in UI. Just need to check it after 1731.

**Comment by D3nnis3n** [comment]
Well, if that was fixed, great. It didnt work at the beginning of Tiger.
All accounts were created with the correct account but money didnt get into
that after restart.

**Comment by mirasrael** [comment]
Seems originally mentioned issue was fixed. For following up comments I suggest to create two separate issues:
- Suggestion: Remember last selected account (for post-9.0)
- Improvement: Show info about target account (may be implemented in the tooltip for 9.0).
